torchtune/data/_messages.py

```diff
@@ -361,11 +380,31 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
             role = role_map[message["from"]]
             if role == "system" and self.new_system_prompt is not None:
                 continue
-            content = message["value"]
+            content = (
+                split_text_by_image_tag(message["value"], self.image_tag)
+                if role == "user" and self.image_tag is not None
+                else message["value"]
+            )
             masked = (role != "assistant") and (not self.train_on_input)
             messages.append(Message(role=role, content=content, masked=masked))
 
-        return {"messages": messages}
+        # Retrieve image from image_dir if specified
+        if "image" in sample and isinstance(sample["image"], str):
+            if self.image_dir is None:
+                raise ValueError(
+                    "You must specify an image_dir to load images specified with paths."
+                )
+            image_path = os.path.join(self.image_dir, sample["image"])
+            image = Image.open(image_path)
+        elif "image" in sample and isinstance(sample["image"], Image.Image):
+            image = sample["image"]
+        else:
+            image = None
+
+        processed_sample = {"messages": messages}
+        if image is not None:
+            processed_sample["images"] = [image]
+        return processed_sample
```

Comment:
Maybe controversial but why do we need to put all this in the existing `ShareGPTToMessages` class? Then as a text-only user I have extra stuff to sort through, plus when we expose at the dataset level it feels kinda overloaded. Idk maybe there are obvious pros to doing it this way that I'm missing here (apart from the obvious one of less code duplication, which tbh I don't care too much about in this case)

Reply:
It seemed general enough to include in a general transform, but yeah I can see this being overloaded. Some options are: 1) make a separate ShareGPT transform with images, 2) keep the llava instruct transform its own class until we find more examples that fit this format (don't generalize until we need to ;) )

Reply:
OK yeah let's do (2). Plus we already have a Cauldron transform so I feel like it's most consistent to do it this way anyways